messagequeue/messagequeue.go

```diff
@@ -0,0 +1,216 @@
+package messagequeue
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	bsmsg "github.com/ipfs/go-bitswap/message"
+	bsnet "github.com/ipfs/go-bitswap/network"
+	wantlist "github.com/ipfs/go-bitswap/wantlist"
+	logging "github.com/ipfs/go-log"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+var log = logging.Logger("bitswap")
+
+type MessageQueue interface {
+	RefIncrement()
+	RefDecrement() bool
+	AddMessage(entries []*bsmsg.Entry, ses uint64)
+	Startup(ctx context.Context, initialEntries []*wantlist.Entry)
+	Shutdown()
+}
+
+type msgQueue struct {
+	p peer.ID
+
+	outlk   sync.Mutex
+	out     bsmsg.BitSwapMessage
+	network bsnet.BitSwapNetwork
+	wl      *wantlist.ThreadSafe
+
+	sender bsnet.MessageSender
+
+	refcnt int
+
+	work chan struct{}
+	done chan struct{}
+}
+
+func New(p peer.ID, network bsnet.BitSwapNetwork) MessageQueue {
+	return &msgQueue{
+		done:    make(chan struct{}),
+		work:    make(chan struct{}, 1),
+		wl:      wantlist.NewThreadSafe(),
+		network: network,
+		p:       p,
+		refcnt:  1,
+	}
+}
+
+func (mq *msgQueue) RefIncrement() {
+	mq.refcnt++
+}
+
+func (mq *msgQueue) RefDecrement() bool {
+	mq.refcnt--
+	return mq.refcnt > 0
+}
+
+func (mq *msgQueue) AddMessage(entries []*bsmsg.Entry, ses uint64) {
```

Comment:
`ses`?

Reply:
session id. I don't make the variable names, I just copy-paste them. :)
But good point and I'll work on some variable names too -- just one at  a time.